File: tests/performance/drivers/python/app/cold_start_execution.py
"""Cold-start child process: measures import → connect → SELECT 1.

Invoked as a subprocess by main.py for each cold-start iteration.
Prints a single CSV row to stdout so the parent can aggregate results.

Environment variables:
    CONNECTION_PARAMS_JSON  – JSON-encoded dict of snowflake.connector.connect() kwargs
    DRIVER_TYPE             – "universal" or "old" (for WireMock TLS/proxy setup)
"""
import json
import os
import resource
import sys
import time

t0 = time.perf_counter()

from snowflake import connector  # noqa: E402  — import timing is the point

t1 = time.perf_counter()

from config import _disable_tls_verification_for_wiremock

_disable_tls_verification_for_wiremock(os.getenv("DRIVER_TYPE", "universal"))

params = json.loads(os.environ["CONNECTION_PARAMS_JSON"])
conn = connector.connect(**params)

t2 = time.perf_counter()

cur = conn.cursor()
cur.execute("SELECT 1")
row = cur.fetchone()
assert row[0] == 1, f"Expected 1, got {row[0]}"
cur.close()
conn.close()

t3 = time.perf_counter()

timestamp_ms = int(time.time() * 1000)
e2e_s = t3 - t0
load_s = t1 - t0
connect_s = t2 - t1
select1_s = t3 - t2
cpu_time_s = time.process_time()
if sys.platform == "darwin":
    peak_rss_mb = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss / (1024 * 1024)
else:
    peak_rss_mb = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss / 1024

print(f"{timestamp_ms},{e2e_s:.6f},{load_s:.6f},{connect_s:.6f},{select1_s:.6f},{cpu_time_s:.6f},{peak_rss_mb:.1f}")

File: tests/performance/drivers/nodejs/app/cold_start_execution.js
'use strict';

// Cold-start child process: measures import -> connect -> SELECT 1.
//
// Invoked as a subprocess by main.js for each cold-start iteration. Prints a
// single CSV row to stdout so the parent can aggregate.
//
// Environment variables:
//   CONNECTION_PARAMS_JSON  - JSON-encoded connection options
//   DRIVER_TYPE             - "universal" or "old" (unused here: WireMock/
//                              recorded_http is not supported for nodejs yet,
//                              so there is no proxy/TLS relaxation to apply)

const t0 = process.hrtime.bigint();

const sdk = require('snowflake-sdk'); // import timing is the point

const t1 = process.hrtime.bigint();

function toSeconds(startNs, endNs) {
  return Number(endNs - startNs) / 1e9;
}

function connectAsync(params) {
  const connection = sdk.createConnection(params);
  return connection.connectAsync().then(() => connection);
}

function executeAsync(connection, sqlText) {
  return new Promise((resolve, reject) => {
    connection.execute({
      sqlText,
      complete: (err, stmt, rows) => {
        if (err) reject(err);
        else resolve(rows);
      },
    });
  });
}

async function main() {
  const connectionParams = JSON.parse(process.env.CONNECTION_PARAMS_JSON);

  const connection = await connectAsync(connectionParams);
  const t2 = process.hrtime.bigint();

  const rows = await executeAsync(connection, 'SELECT 1');
  const value = rows && rows.length > 0 ? Object.values(rows[0])[0] : undefined;
  if (Number(value) !== 1) {
    throw new Error(`Expected 1, got ${value}`);
  }

  const t3 = process.hrtime.bigint();

  const timestampMs = Date.now();
  const e2eS = toSeconds(t0, t3);
  const loadS = toSeconds(t0, t1);
  const connectS = toSeconds(t1, t2);
  const select1S = toSeconds(t2, t3);
  const cpuUsage = process.cpuUsage();
  const cpuTimeS = (cpuUsage.user + cpuUsage.system) / 1e6;
  const peakRssMb = process.resourceUsage().maxRSS / 1024;

  process.stdout.write(
    `${timestampMs},${e2eS.toFixed(6)},${loadS.toFixed(6)},${connectS.toFixed(6)},${select1S.toFixed(6)},${cpuTimeS.toFixed(6)},${peakRssMb.toFixed(1)}\n`,
  );
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
